File: yoshimoto/poly_fit_manual.cpp
#include <TGraph.h>
#include <TLine.h>
#include <TLegend.h>
#include <vector>
#include <fstream>
#include <iostream>
using namespace std;

void poly_fit_manual(bool fit=false)
{
    vector<double> vx, vy;

    ifstream ifs("temperature.txt");
    double temp;
    int i = 1;
    while (ifs >> temp)
    {
        vx.push_back(i++);
        vy.push_back(temp);
    }

    int N = vx.size();

    TGraph *graph = new TGraph(N, &(vx.at(0)), &(vy.at(0)));
    graph->SetTitle(";Time [day];Body temperature [#circ C]");
    graph->SetMinimum(36);
    graph->SetMaximum(38);
    graph->SetMarkerStyle(2);
    graph->Draw("AP");

    if(fit == false){
        TLegend *legend = new TLegend(0.15,0.80,0.3,0.85);
        legend->AddEntry(graph,"Data","p");
        legend->Draw();
        return;
    }

    // 頻出する総和を先に計算
    double Nxy = 0, Nx = 0, Ny = 0, Nx2 = 0;
    for (int i = 0; i < N; i++)
    {
        const double xi = vx[i];
        const double yi = vy[i];
        Nxy += xi * yi;
        Nx += xi;
        Ny += yi;
        Nx2 += xi * xi;
    }
    double delta = N * Nx2 - Nx * Nx;

    // 係数
    double b = (N  * Nxy - Nx * Ny ) / delta;
    double a = (Nx2 * Ny - Nx * Nxy) / delta;

    // Yの誤差
    double dy = 0;
    for (int i = 0; i < N; i++)
    {
        const double xi = vx[i];
        const double yi = vy[i];
        dy += (b * xi + a - yi) * (b * xi + a - yi);
    }
    double sigma_y = sqrt(1.0 / (N - 2) * dy);

    // 係数の誤差
    double sigma_a = sigma_y * sqrt(Nx2 / delta);
    double sigma_b = sigma_y * sqrt(N / delta);

    // パラメータの出力
    cout << "N      : " << N << endl;
    cout << "A      : " << a << "+/-" << sigma_a << endl;
    cout << "B      : " << b << "+/-" << sigma_b << endl;
    cout << "Sigma_y: " << sigma_y << endl;

    // グラフ生成
    TLine *line = new TLine(0, a + 0 * b, 70, a + 70 * b );
    line->SetLineColor(kRed);
    line->Draw();

    if(fit){
        TLegend *legend = new TLegend(0.15,0.75,0.3,0.85);
        legend->AddEntry(graph,"Data","p");
        legend->AddEntry(line,"Fitted line","l");
        legend->Draw();
    }
}

File: yoshimoto/poly_fit_manual.py
import numpy as np
import matplotlib.pyplot as plt
import sys
import my_style

def poly_fit_manual(fit=False):
    vx=[]
    vy=[]
    i=1
    with open("temperature.txt") as f:
        for line in f:
            vx.append(i)
            vy.append(float(line.split()[0]))
            i+=1

    # グラフ生成
    plt.plot(vx,vy,'k+',label='Data')
    plt.xlabel("Time [day]")
    plt.ylabel("Body temperature [$^\\circ$C]")
    plt.ylim(36,38)
    
    if fit==False:
        plt.legend()
        plt.show()
        return


    # 頻出する総和を先に計算
    N   = len(vy)
    Nxy = np.sum([xi*yi for xi,yi in zip(vx,vy)])
    Nx  = np.sum([xi    for xi,yi in zip(vx,vy)])
    Ny  = np.sum([yi    for xi,yi in zip(vx,vy)])
    Nx2 = np.sum([xi*xi for xi,yi in zip(vx,vy)])
    delta=N * Nx2 - Nx**2

    # 係数
    a = (Nx2 * Ny  - Nx * Nxy)/delta
    b = (N   * Nxy - Nx * Ny )/delta

    # Yの誤差
    sigma_y = np.sqrt(1/(N-2) * np.sum([(a+b*xi-yi)**2 for xi,yi in zip(vx,vy)]))

    # 係数の誤差
    sigma_a = sigma_y * np.sqrt(Nx2 / delta)
    sigma_b = sigma_y * np.sqrt(N   / delta)

    # パラメータの出力
    print(f"N      : {N}")
    print(f"A      : {a:7f} ± {sigma_a:7f}")
    print(f"B      : {b:7f} ± {sigma_b:7f}")
    print(f"Sigma_y: {sigma_y:7f}")

    # グラフ生成
    xx=np.linspace(0,70,100)
    plt.plot(xx,b*xx+a,label="Fitted line",color="red")

    plt.legend()
    plt.show()

if __name__ == "__main__":
    if len(sys.argv) == 1:
        poly_fit_manual()
    if len(sys.argv) == 2:
        poly_fit_manual(bool(sys.argv[1]))
